src/libmw/src/node/BlockBuilder.cpp

```diff
@@ -0,0 +1,97 @@
+#include <mw/node/BlockBuilder.h>
+#include <mw/consensus/KernelSumValidator.h>
+#include <mw/consensus/Aggregation.h>
+#include <mw/consensus/Weight.h>
+
+#include <unordered_set>
+#include <numeric>
+
+MW_NAMESPACE
+
+bool BlockBuilder::AddTransaction(const Transaction::CPtr& pTransaction, const std::vector<PegInCoin>& pegins)
+{
+    // Check weight
+    uint64_t weight = Weight::Calculate(pTransaction->GetBody());
+    if ((weight + m_weight) > libmw::MAX_BLOCK_WEIGHT) {
+        LOG_ERROR("Exceeds max block weight");
+        return false;
+    }
+    
+    // Verify pegin amount matches
+    const uint64_t actual_amount = pTransaction->GetPegInAmount();
+    const uint64_t expected_amount = std::accumulate(pegins.cbegin(), pegins.cend(), (uint64_t)0,
+        [](const uint64_t sum, const PegInCoin& pegin) { return sum + pegin.GetAmount(); }
+    );
+    if (actual_amount != expected_amount) {
+        LOG_ERROR("Mismatched pegin amount");
+        return false;
+    }
+
+    // Verify pegin commitments are unique
```

Comment:
Can share the same code as the duplicate commit check in TxBody::Validate

Reply:
This might also be a redundant check if we assume pegins is a redundant parameter, assuming kernel commits are always unique.